ksqldb-execution/src/main/java/io/confluent/ksql/execution/util/FunctionArgumentsUtil.java

```diff
@@ -0,0 +1,227 @@
+   * Function(U,R)).
+   * In the second pass, we use the LambdaType inputTypes field to construct SqlLambdaResolved
+   * that has the proper input type list and return type. We also need to construct a list of
+   * contexts that should be used when processing each function argument subtree.
+   *
+   * @param expressionTypeManager an expression type manager
+   * @param functionCall  the function expression
+   * @param udfFactory  udf factory for the function in the expression
+   * @param context a type context
+   *
+   * @return a wrapper that contains a list of function arguments 
+   *         (any lambdas are SqlLambdaResolved), the ksql function, 
+   *         type contexts for use in further processing the function 
+   *         argument child nodes, and the return type of the ksql function
+   */
+  // CHECKSTYLE_RULES.OFF: CyclomaticComplexity
+  public static FunctionArgumentsAndContext getLambdaContextAndType(
+      final ExpressionTypeManager expressionTypeManager,
+      final FunctionCall functionCall,
+      final UdfFactory udfFactory,
+      final TypeContext context
+  ) {
+    // CHECKSTYLE_RULES.ON: CyclomaticComplexity
+    final List<Expression> arguments = functionCall.getArguments();
+    final List<SqlArgument> functionArgumentTypes = firstPassOverFunctionArguments(
+        arguments,
+        expressionTypeManager,
+        context
+    );
+
+    final KsqlScalarFunction function = udfFactory.getFunction(functionArgumentTypes);
+    final SqlType returnSchema;
+    final List<TypeContext> typeContextsForChildren = new ArrayList<>();
+
+    if (!functionCall.hasLambdaFunctionCallArguments()) {
+      returnSchema = function.getReturnType(functionArgumentTypes);
+      arguments.forEach(argument -> typeContextsForChildren.add(context.getCopy()));
+      return new FunctionArgumentsAndContext(
+          typeContextsForChildren,
+          functionArgumentTypes,
+          returnSchema,
+          function
+      );
+    } else {
+      final List<ParamType> paramTypes = function.parameters();
+      final Map<GenericType, SqlType> reservedGenerics = new HashMap<>();
+      final List<SqlArgument> functionArgumentTypesWithResolvedLambdaType = new ArrayList<>();
+
+      // second pass over the function arguments to properly do lambda type checking
+      for (int i = 0; i < arguments.size(); i++) {
+        final Expression expression = arguments.get(i);
+        final ParamType parameter = paramTypes.get(i);
+        if (expression instanceof LambdaFunctionCall) {
+          final TypeContext childContext = context.getCopy();
+
+          // the function returned from the UDF factory should have lambda
+          // at this index in the function arguments if there's a
+          // lambda node at this index in the function node argument list
+          if (!(parameter instanceof LambdaType)) {
+            throw new RuntimeException("Error while processing lambda function. "
```

Comment:
nit: maybe print the `ParamType` in the error message as well?

Reply:
done